src/fsharp/NameResolution.fs

```diff
@@ -763,26 +763,81 @@ let AddStaticContentOfTyconRefToNameEnv (g:TcGlobals) (amap: Import.ImportMap) m
-
-               for (methName, methGroup) in methGroups do
-                   let methGroup = methGroup |> List.filter (fun m -> not m.IsInstance && not m.IsClassConstructor)
-                   if not methGroup.IsEmpty then
-                       yield KeyValuePair(methName, Item.MethodGroup(methName, methGroup, None)) 
-           
-               let propInfos = 
-                   AllPropInfosOfTypeInScope ResultCollectionSettings.AllResults infoReader nenv None AccessorDomain.AccessibleFromSomeFSharpCode PreferOverrides m ty
-                   |> List.groupBy (fun m -> m.PropertyName)
-               
-               for (propName, propInfos) in propInfos do
-                   let propInfos = propInfos |> List.filter (fun m -> m.IsStatic)
-                   for propInfo in propInfos do 
-                       yield KeyValuePair(propName , Item.Property(propName,[propInfo])) |]
-
-        { nenv with eUnqualifiedItems = nenv.eUnqualifiedItems.AddAndMarkAsCollapsible items }
+            // This is to handle the current behavior with qualified extension members:
+            //     Extension member properties will hide all extension methods that share the same name.
+            [|
+                let propItems = getPropItems ()
+                let methItems = getMethItems ()
+
+                let itemGroups =
+                    propItems @ methItems
+                    |> List.groupBy (fun pair -> pair.Key)
+
+                for (_, items) in itemGroups do
+                    let sortedItems =
+                        items 
+                        |> List.sortBy (fun pair -> 
+                            match pair.Value with 
+                            | Item.Property (_, [propInfo]) -> not propInfo.IsExtensionMember 
+                            | Item.MethodGroup (_, methGroup, _) -> methGroup |> List.exists (fun x -> not x.IsExtensionMember)
+                            | _ -> false)
+                    match sortedItems with
+                    | item :: _ -> yield item
+                    | _ -> ()
+            |]
+
+        { nenv with 
+            eUnqualifiedItems = 
+                (nenv.eUnqualifiedItems, items) 
+                ||> Array.fold (fun x (KeyValue(k, v)) -> 
+                    match x.TryGetValue k with
+                    | true, existingItem ->
+                        match existingItem, v with
+                        
+                        // Combine methods groups.
+                        | Item.MethodGroup (_, minfosCurrent, None), Item.MethodGroup (_, minfosNew, None) ->
+                            let minfos =
+                                (minfosNew, minfosCurrent)
+                                ||> List.fold (fun minfos minfo1 ->
+                                    // Shadow methods with the same signature.
+                                    if minfos |> List.exists (fun minfo2 -> MethInfosEquivByNameAndSig Erasure.EraseAll true g amap m minfo1 minfo2) then
```

Comment:
This looks quadratic in the number of method overloads?  I suspect it's hard to escape that without massively over-complicating the code though. Other things in the compiler are also quadratic (though I think it's less common to be quadratic when consuming a construct.)

Reply:
We should consider whether we should be using `EraseAll` here - e.g. should units of measure be significant when determining whether one overload shadows another? EraseAll means they are not. I think they _should_ be significant - because the only reason we erase UoM when forbidding identical method signatures is that .NET IL doesn't permit two identical method signatures within the same class.  But it does allow them across different classes.

Either way we will definitely need test cases for this case

Reply:
I think I can make a slight tweak to the overloads complexity; I notice I keep searching on the list that keeps getting additions which isn't necessary.

In regards to `EraseAll`. Because you cannot do overloading with UoM normally in classes, should we now allow it in this case? I was worried about consistency. **But**, it would be extremely powerful to allow it. I'll set it back to `EraseNone`; it should be pretty interesting to see overloading with UoM. I can imagine reasonable use cases where it would be helpful. :)

Reply:
It's a tricky choice - each way sort of seems wrong. Could you try the switch and explore some test cases?